File: analyzers/tests/SonarAnalyzer.Test/TestCases/AvoidUnsealedAttributes.vb
' Test framework can't assert location on 1st line
Public Class PublicAttribute ' Noncompliant {{Seal this attribute or make it abstract.}}
    Inherits Attribute
    '        ^^^^^^^^^^^^^^^ @-1
End Class

Public NotInheritable Class SealedAttribute
    Inherits Attribute ' Compliant
End Class

Public MustInherit Class AbstractAttribute
    Inherits Attribute ' Compliant
End Class

Public Class NotAnAttribute ' Compliant - not an attribute
End Class

Public Class Container

    Protected Class ProtectedAttribute ' Noncompliant
        Inherits Attribute
    End Class

    Private Class PrivateAttribute ' Compliant
        Inherits Attribute

        Public Class SubClassAttribute ' Compliant - effective accessibility is private
            Inherits Attribute
        End Class
    End Class
End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AvoidUnsealedAttributes.cs
using System;

public class MyAttribute : Attribute { } // Noncompliant {{Seal this attribute or make it abstract.}}
//           ^^^^^^^^^^^

public class MyOtherAttribute : Attribute { } // Noncompliant

public sealed class Bar : Attribute { } // Compliant - sealed


public abstract class FooBar : Attribute { } // Compliant - abstract

public class NotAnAttribute { } // Compliant - not an attribute

public sealed class Attr : Attribute
{
    private class InnerAttr : Attribute // Compliant - private
    {
        public class InnerInnerAttr : Attribute { } // Compliant - effective accessibility is private
    }

    protected class InnerAttr2 : Attribute { } // Noncompliant
}

public class { }    // Error [CS1001]  Identifier expected
